File: better-ddd/src/test/java/com/example/AggregateRootConstructorAspect.java
package com.example;

import org.aspectj.lang.JoinPoint;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Around;
import org.aspectj.lang.annotation.Aspect;
import org.aspectj.lang.annotation.Before;
import org.aspectj.lang.annotation.Pointcut;

import java.util.Arrays;

@Aspect
public class AggregateRootConstructorAspect {

    @Pointcut("execution(*.new(..)) && @within(org.morecup.jimmerddd.betterddd.annotation.AggregateRoot)")
    public void constructorOfAggregateRoot() {}

    @Before("constructorOfAggregateRoot()")
    public void beforeConstructor(JoinPoint joinPoint) {
        System.out.println("即将构造 AggregateRoot 对象：" + joinPoint.getSignature());
    }

    // 如果想用 @Around 拦截，注意构造函数无返回值，不能替换实例
    @Around("constructorOfAggregateRoot()")
    public Object aroundConstructor(ProceedingJoinPoint pjp) throws Throwable {
        System.out.println("环绕通知：构造 AggregateRoot 对象，参数：" + Arrays.toString(pjp.getArgs()));

        Object obj = pjp.proceed();

        System.out.println("构造完成：" + obj); // 这里 obj 是 null，因为构造函数无返回值

        return obj;
    }

//    @Pointcut("call(*.new(..)) && @within(org.morecup.jimmerddd.betterddd.annotation.AggregateRoot)")
//    public void callConstructorOfAggregateRoot() {}
}

File: better-ddd-aspect/src/main/kotlin/org/morecup/jimmerddd/betterddd/aspect/AggregateRootConstructorAspect.kt
package org.morecup.jimmerddd.betterddd.aspect

import org.aspectj.lang.JoinPoint
import org.aspectj.lang.ProceedingJoinPoint
import org.aspectj.lang.annotation.Aspect
import org.aspectj.lang.annotation.Before
import org.aspectj.lang.annotation.Pointcut
import org.morecup.jimmerddd.betterddd.core.bridge.ConstructorBridgeConfig

@Aspect
class AggregateRootConstructorAspect {

    @Pointcut("execution(*.new(..)) && @within(org.morecup.jimmerddd.betterddd.core.annotation.AggregateRoot)")
    fun constructorOfAggregateRoot() {}

    @Before("constructorOfAggregateRoot()")
    fun beforeConstructor(joinPoint: JoinPoint) {
        ConstructorBridgeConfig.constructorBridge.createInstance(joinPoint as ProceedingJoinPoint, joinPoint.args)
//        println("About to construct AggregateRoot object: ${joinPoint.signature}")

    }

    // If using @Around to intercept, note that constructors don't have return values, so you can't replace the instance
//    @Around("constructorOfAggregateRoot()")
//    fun aroundConstructor(pjp: ProceedingJoinPoint): Any? {
//        println("Around advice: Constructing AggregateRoot object, arguments: ${pjp.args.contentToString()}")
//
//        val obj = pjp.proceed()
//
//        println("Construction completed: $obj") // obj is null here because constructors don't return values
//
//        return obj
//    }

//    @Pointcut("call(*.new(..)) && @within(org.morecup.jimmerddd.betterddd.annotation.AggregateRoot)")
//    fun callConstructorOfAggregateRoot() {}
}